File: HtProject-sp-callback/src/com/database/Logical/LightDataModel.java
package com.database.Logical;

import java.lang.reflect.Method;
import java.lang.reflect.Type;
import java.security.KeyException;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * 基础Model驱动
 * 
 * @author Shotgun
 *
 */
public abstract class LightDataModel implements com.database.Interface.IUpdateInfo {

	/**
	 * 有变化的字段,通常表示需要更新到数据库
	 */
	ArrayList<String> updatedFields;

	/**
	 * 存储当前值为空的字段名
	 */
	ArrayList<String> NullFields;
	/**
	 * 存储未知的字段值
	 */
	HashMap<String, Object> extrFields;
	/**
	 * 是否正在填充数据
	 */
	Boolean _dataFilling = false;
	/**
	 * 是否要忽略相同的更新
	 */
	Boolean _IgnoreEquals = false;

	public LightDataModel() {
		updatedFields = new ArrayList<String>();
		extrFields = new HashMap<String, Object>();

	}

	private Method GetMethod(String mName) {
		Class<? extends LightDataModel> cls = this.getClass();
		Method[] ms = cls.getMethods();
		// System.out.println("get method:"+ mName);
		for (Method m : ms) {
			if (m.getName().equalsIgnoreCase(mName)) {
				return m;
			}
		}
		// System.out.println(".... not found");
		return null;
	}

	/**
	 * 通过反射获取成员成公用属性值
	 * 
	 * @param key
	 *            成员名
	 * @param obj
	 *            取得的成员值
	 * @return 是否成功
	 */
	public ReturnObj GetValue(String key) {
		Method func = this.GetMethod("get_" + key);
		ReturnObj ret = new ReturnObj();
		ret.IsSucces = false;
		if (func == null)
			return ret;
		try {
			ret.Value = func.invoke(this);
			ret.IsSucces = true;
		} catch (Exception e) {
			e.printStackTrace();
		}
		return ret;
	}

	/**
	 * 通过反射设置成员成公用属性值
	 * 
	 * @param key
	 *            成员名
	 * @param value
	 *            设置的成员值
	 * @return
	 */
	public Boolean SetValue(String key, Object value) {
		Method func = GetMethod("set_" + key);
		if (func == null)
			return false;
		boolean iToDefault = value == null;
		String targetType = null;
		if (iToDefault) {
			Type[] ts = func.getGenericParameterTypes();
			targetType = ts[0].getTypeName();

			if ("byte".equals(targetType) || "int".equals(targetType) || "long".equals(targetType)
					|| "short".equals(targetType)) {
				value = 0;
			} else if ("float".equals(targetType))
				value = 0f;
			else if ("double".equals(targetType))
				value = 0.0;
			else if ("boolean".equals(targetType))
				value = false;
			else if ("char".equals(targetType))
				value = (char) 0;
			else
				iToDefault = false;
		}

		try {
			// System.out.println("setValue " + key + "=" + value);
			func.invoke(this, value);
		} catch (

		Exception e)

		{
			if (iToDefault)
				System.out.println("set  key " + key + " default value fail,target type " + targetType + ",error:"
						+ e.getMessage());
			else {
				String valueType = "null";
				if (value != null) {
					valueType = value.getClass().getTypeName();
				}
				System.out.println(
						"set  key " + key + " value fail, value type:" + valueType + ",error:" + e.getMessage());
			}
			return false;
		}
		if (iToDefault)

		{
			SetNullFlag(key);
		}
		return true;

	}

	public Boolean IsNull(String Field) {
		if (this.NullFields == null)
			return false;
		return this.NullFields.contains(Field.toLowerCase());
	}

	/**
	 * 将字段值设置为null
	 */
	public void SetNull(String key) {// throws KeyException {
		// 引用类型可以此处将被标记为null,值类型无效
		if (!SetValue(key, null)) {
			System.out.println(String.format("Key: %s not Found", key));
			// throw new KeyException(String.format("Key: %s not Found", key));
		}
		SetNullFlag(key);
	}

	/**
	 * 设置段值空标记
	 * 
	 * @param key
	 * @throws KeyException
	 *             设置不允许为空的字段
	 */
	protected void SetNullFlag(String key) { // throws KeyException
		if (IsNull(key))
			return;
		String[] fs = GetNullableFields();
		Boolean Exist = false;
		for (String s : fs) {
			if (key.equalsIgnoreCase(s)) {
				Exist = true;
				break;
			}
		}

		if (!Exist && !this.GetDataFilling()) {
			System.out.println(String.format("Key:%s Non-Nullable ", key));
			// throw new KeyException(String.format("Key:%s Non-Nullable ",
			// key));
		}
		if (NullFields == null)
			NullFields = new ArrayList<String>();
		this.NullFields.add(key.toLowerCase());
	}

	/**
	 * 删除去字段null标记
	 */
	protected void RemoveNullFlag(String key) {
		if (NullFields == null)
			return;
		String lk = key.toLowerCase();
		if (NullFields.contains(lk))
			NullFields.remove(lk);
	}

	protected abstract String[] GetNullableFields();

	/**
	 * 是否正在填充数据
	 * 
	 * @return
	 */
	Boolean GetDataFilling() {
		return this._dataFilling;
	}

	/**
	 * 设置当前是否正在填充数据
	 */
	void SetDataFilling(Boolean value) {
		this._dataFilling = value;
	}

	protected void SetFieldHasUpdate(String field, Object oVal, Object nVal) {
		if (this.GetDataFilling())
			return;
		if (this.GetIgnoreEquals()) {
			if (oVal == null) {
				if (nVal == null)
					return;
			} else if (oVal.equals(nVal))
				return;
		}
		if (updatedFields.contains(field))
			return;
		updatedFields.add(field);
	}

	/**
	 * 赋值时检查是不是否与原值相同，相同时将忽略该操作(在执行存入数据库时，该字段不会有任何表现)
	 */
	public Boolean GetIgnoreEquals() {
		return this._IgnoreEquals;
	}

	/**
	 * 赋值时检查是不是否与原值相同，相同时将忽略该操作(在执行存入数据库时，该字段不会有任何表现)
	 */
	public void SetIgnoreEquals(Boolean value) {
		this._IgnoreEquals = value;
	}

	@Override
	public String[] GetUpateFields() {
		if (this.updatedFields == null || this.updatedFields.size() == 0)
			return null;
		return this.updatedFields.toArray(new String[0]);
	}

	@Override
	public Object GetValueByName(String fields) {
		ReturnObj ret = this.GetValue(fields);
		if (ret.IsSucces)
			return ret.Value;
		return false;
	}

	@Override
	public void SetUpdated(Object IdentifyValue) {
		this.updatedFields.clear();
		if (IdentifyValue == null)
			return;
		if (IdentifyValue instanceof Long) {
			IdentifyValue = ((Long) IdentifyValue).intValue();
			// System.out.println("long to int");
		}
		this.SetValue(this.IdentifyField(), IdentifyValue);
		// if (IdentifyValue is decimal)
		// this[IdentifyField] = Convert.ToInt32((decimal)IdentifyValue);
		// else
		// this[IdentifyField] = IdentifyValue;

	}

	@Override
	public String Schema() {
		return null;
	}

	public abstract String TableName();

	public abstract String IdentifyField();

}

File: xtone-dotnet-interface/Shotgun.Library/Model/Logical/LightDataModel.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace Shotgun.Model.Logical
{
    /// <summary>
    /// 轻质量数据模块，用于替换原旧的BaseDataRow对像
    /// </summary>
    public abstract class LightDataModel : Shotgun.Database.IUpatedataInfo
    {
        /// <summary>
        /// 表示已经更新数据
        /// </summary>
        List<string> updatedFields;

        /// <summary>
        /// 为空的字段,注意转为小写
        /// </summary>
        List<string> NullFields;

        /// <summary>
        /// 用于存储未知的外部字段
        /// </summary>
        Dictionary<string, object> extrFields;

        public LightDataModel()
        {
            updatedFields = new List<string>();
            extrFields = new Dictionary<string, object>();
        }

        #region IUpatedataInfo 成员

        List<string> Database.IUpatedataInfo.GetUpateFields()
        {
            return updatedFields;
        }

        object Database.IUpatedataInfo.GetValueByName(string filds)
        {
            object obj;
            if (GetValue(filds, out obj))
                return obj;
            throw new KeyNotFoundException(string.Format("Key: {0} not Found", filds));
        }

        void Database.IUpatedataInfo.SetUpdated(object IdentifyValue)
        {
            updatedFields.Clear();

            if (IdentifyValue == null)
            {
                OnSaved(false);
                return;
            }
            if (IdentifyValue is decimal)
                this[IdentifyField] = Convert.ToInt32((decimal)IdentifyValue);
            else
                this[IdentifyField] = IdentifyValue;
            OnSaved(true);
        }

        public abstract string IdentifyField { get; }

        public abstract string TableName { get; }

        #endregion

        /// <summary>
        /// 将数据写入数据库
        /// </summary>
        /// <param name="dBase"></param>
        /// <returns></returns>
        public bool SaveToDatabase(Shotgun.Database.IBaseDataClass2 dBase)
        {
            return dBase.SaveData(this);
        }

        protected virtual void OnSaved(bool isInsert) { }

        /// <summary>
        /// 存取字段值,字段值为null时返回default(type);存入null,会调用SetNull(key)
        /// </summary>
        /// <param name="key"></param>
        /// <returns></returns>
        public object this[string key]
        {
            get
            {
                if (extrFields.ContainsKey(key))
                    return extrFields[key];
                object obj;
                if (GetValue(key, out obj))
                    return obj;
                throw new KeyNotFoundException(string.Format("Key: {0} not Found", key));
            }
            set
            {
                if (value is DBNull)
                    value = null;
                if (extrFields.ContainsKey(key))
                {
                    extrFields[key] = value;
                    return;
                }

                if (SetValue(key, value))
                {
                    if (value == null)//SetValue对值类型无效
                        SetNullFlag(key);
                    return;
                }
                extrFields.Add(key, value);
            }
        }

        /// <summary>
        /// 通过反射获取成员成公用属性值
        /// </summary>
        /// <param name="key">成员名</param>
        /// <param name="obj">取得的成员值</param>
        /// <returns></returns>
        private bool GetValue(string key, out object obj)
        {
            obj = null;
            if (IsNull(key))
                return true;
            var t = this.GetType();
            var p = t.GetProperty(key,
                System.Reflection.BindingFlags.Instance |
                System.Reflection.BindingFlags.Public |
                System.Reflection.BindingFlags.IgnoreCase |
                System.Reflection.BindingFlags.GetProperty);
            if (p == null)
                return false;
            obj = p.GetValue(this, null);
            return true;
        }

        /// <summary>
        /// 通过反射设置成员成公用属性值
        /// </summary>
        /// <param name="key">成员名</param>
        /// <param name="value">设置的成员值</param>
        /// <returns></returns>
        private bool SetValue(string key, object value)
        {
            var t = this.GetType();
            var p = t.GetProperty(key,
                System.Reflection.BindingFlags.Instance |
                System.Reflection.BindingFlags.Public |
                System.Reflection.BindingFlags.IgnoreCase |
                System.Reflection.BindingFlags.GetProperty);
            if (p == null)
                return false;
            if (p.PropertyType.Name.Equals("Boolean"))
            {
                if (!(value is bool))
                    value = Convert.ToBoolean(value);
                //if (p.MemberType ==typeof( Boolean)
            }
            p.SetValue(this, value, null);
            return true;
        }

        public bool IsNull(string key)
        {
            if (this.NullFields == null)
                return false;
            return this.NullFields.Contains(key.ToLower());
        }

        /// <summary>
        /// 将字段值设置为null
        /// </summary>
        public void SetNull(string key)
        {
            if (!SetValue(key, null)) //引用类型可以此处将被标记为null,值类型无效
                throw new KeyNotFoundException(string.Format("Key: {0} not Found", key));
            SetNullFlag(key);
        }

        /// <summary>
        /// 设置段值空标记
        /// </summary>
        /// <param name="key"></param>
        protected void SetNullFlag(string key)
        {
            if (IsNull(key))
                return;
            var fs = GetNullableFields();
            bool Exist = false;
            foreach (var s in fs)
            {
                if (key.Equals(s, StringComparison.OrdinalIgnoreCase))
                {
                    Exist = true;
                    break;
                }
            }

            if (!Exist && !DataFilling)
                throw new ArgumentException(string.Format("Key:{0} Non-Nullable ", key));
            if (NullFields == null)
                NullFields = new List<string>();
            this.NullFields.Add(key.ToLower());
        }

        /// <summary>
        /// 删除去字段null标记
        /// </summary>
        /// <param name="key"></param>
        protected void RemoveNullFlag(string key)
        {
            if (NullFields == null)
                return;
            var lk = key.ToLower();
            if (NullFields.Contains(lk))
                NullFields.Remove(lk);
        }

        /// <summary>
        /// 取得可以为空的字段
        /// </summary>
        /// <returns></returns>
        protected abstract string[] GetNullableFields();

        protected void SetFieldHasUpdate(string field, object oVal, object nVal)
        {
            if (DataFilling)
                return;
            if (this.IgnoreEquals)
            {
                if (oVal == null)
                {
                    if (nVal == null)
                        return;
                }
                else if (oVal.Equals(nVal))
                    return;
            }
            if (updatedFields.Contains(field))
                return;
            updatedFields.Add(field);
        }

        /// <summary>
        /// 赋值时检查是不是否与原值相同，相同时将忽略该操作(在执行存入数据库时，该字段不会有任何表现)
        /// </summary>
        public bool IgnoreEquals { get; set; }

        /// <summary>
        /// 表示是否正在填充数据
        /// </summary>
        internal bool DataFilling { get; set; }

        /// <summary>
        /// 将数据复制到DataTable中
        /// 主键值已经存在将抛出ArgumentException
        /// </summary>
        /// <param name="dt"></param>
        /// <returns>复制的新行</returns>
        public System.Data.DataRow CopyToDataTable(System.Data.DataTable dt)
        {
            System.Data.DataRow row;
            if (dt.Rows.Count > 0 && dt.PrimaryKey.Length > 0)
            {
                row = dt.Rows.Find(new object[] { this[dt.PrimaryKey[0].ColumnName] });
                if (row != null)
                    throw new ArgumentException("Key值已经存在");
            }
            row = dt.NewRow();
            object obj;
            foreach (System.Data.DataColumn col in dt.Columns)
            {
                obj = this[col.ColumnName];
                if (obj == null)
                    row[col] = DBNull.Value;
                else
                    row[col] = this[col.ColumnName];
            }
            dt.Rows.Add(row);
            return row;
        }

        /// <summary>
        /// 跨库操作时，需要指定
        /// </summary>
        public virtual string Schema
        {
            get;
            protected set;
        }
    }
}
